Adding kwargs to composable AC API to enable full capabilities

sanketpurandare opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #128516

Summary:
Firstly, this does not change any existing behaviour, since all the
default values for kwargs were hardcoded into the ``_checkpoint_without_reentrant_generator`` call.

Secondly, this is needed for unlocking the full potential of composable
checkpointing making it equivalent to ``torch.utils.checkpoint.checkpoint(use_reentrant=False)``.

Finally, an added benefit is now composable checkpointing can be used under ``FakeTensorMode`` by
passing ``preserve_rng_state=False``.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

sanketpurandare committed:
Adding preserve_rng_state as an arg to composable AC API

Summary:
Firstly, this does not change any existing behaviour, since ``True`` was
hardcoded into the ``_checkpoint_without_reentrant_generator`` call.

Secondly, this is needed for running ``FakeTensor`` through AC, since
rng state preservation and restoration fails in ``FakeTensorMode``.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

sanketpurandare committed:
Update on "Adding preserve_rng_state as an arg to composable AC API"


Summary:
Firstly, this does not change any existing behaviour, since ``True`` was
hardcoded into the ``_checkpoint_without_reentrant_generator`` call.

Secondly, this is needed for running ``FakeTensor`` through AC, since
rng state preservation and restoration fails in ``FakeTensorMode``.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

sanketpurandare committed:
Update on "Adding preserve_rng_state as an arg to composable AC API"


Summary:
Firstly, this does not change any existing behaviour, since ``True`` was
hardcoded into the ``_checkpoint_without_reentrant_generator`` call.

Secondly, this is needed for running ``FakeTensor`` through AC, since
rng state preservation and restoration fails in ``FakeTensorMode``.

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

awgu commented on torch/distributed/_composable/checkpoint_activation.py:
```diff
@@ -13,21 +13,22 @@
 
 @contextmanager
-def _no_hook(module: nn.Module):
+def _no_hook(module: nn.Module, user_ctx: ContextManager = nullcontext()):
```
Note (mainly to myself): This `nullcontext()` object will be constructed at function definition time and reused for all invocations. We are assuming that we do not modify this object, so having it as a default argument is okay. (Otherwise, we would want `user_ctx  = None` and then inside the function, `user_ctx = nullcontext() if user_ctx is None else user_ctx`.)

sanketpurandare replied:
@awgu Although ``nullcontext()`` is initialized at function definition time. We initialize``_no_hook`` everytime in the pre-fw hook. So for any invocation if ``user_ctx is not None``, then ``user_ctx`` will be used.  

But I do see the safety is using what you suggested.

awgu reviewed (commented):
This looks great to me! Do you want to add a unit test :)

awgu commented:
https://github.com/pytorch/pytorch/blob/main/test/distributed/_composable/test_checkpoint.py

Maybe add a unit test to make sure that we can pass `kwargs` through `_composable.checkpoint`?

sanketpurandare committed:
Update on "Adding kwargs to composable AC API to enable full capabilities"


Summary:
Firstly, this does not change any existing behaviour, since all the
default values for kwargs were hardcoded into the ``_checkpoint_without_reentrant_generator`` call.

Secondly, this is needed for unlocking the full potential of composable
checkpointing making it equivalent to ``torch.utils.checkpoint.checkpoint(use_reentrant=False)``.

Finally, an added benefit is now composable checkpointing can be used under ``FakeTensorMode`` by
passing ``preserve_rng_state=False``.

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

sanketpurandare committed:
Update on "Adding kwargs to composable AC API to enable full capabilities"


Summary:
Firstly, this does not change any existing behaviour, since all the
default values for kwargs were hardcoded into the ``_checkpoint_without_reentrant_generator`` call.

Secondly, this is needed for unlocking the full potential of composable
checkpointing making it equivalent to ``torch.utils.checkpoint.checkpoint(use_reentrant=False)``.

Finally, an added benefit is now composable checkpointing can be used under ``FakeTensorMode`` by
passing ``preserve_rng_state=False``.

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

awgu commented on torch/distributed/_composable/checkpoint_activation.py:
```diff
@@ -61,16 +63,45 @@ def checkpoint(module: nn.Module) -> nn.Module:
     torch._C._log_api_usage_once("torch.distributed.checkpoint")
 
-    def forward_pre_hook(module: nn.Module, inputs: Tuple[Any, ...]) -> None:
+    kwargs = kwargs or {}
```
nit: I think we do not need this; if you have `**kwargs` in the signature, then Python will make sure `kwargs` is `{}` if nothing is passed. I think for things like `__torch_dispatch__` we need this because `kwargs` defaults to `None`.

sanketpurandare replied:
Ohh thanks! I didn't know this. Btw why does it default to `None` for `__torch_dispatch__`?

awgu replied:
> Btw why does it default to None for __torch_dispatch__?

I am not too sure. It might be somehow related to the arg parsing logic.

awgu commented on test/distributed/_composable/test_checkpoint.py:
```diff
@@ -219,6 +221,112 @@ def forward(self, x):
+        m = MyModel(True, False)
+        m0, m1, m2, m3 = (deepcopy(m) for _ in range(4))
+
+        # composable checkpoint does not support use_reentrant=True
+        with self.assertRaises(NotImplementedError):
+            checkpoint(m, use_reentrant=True)
+
+        # check giving an unsupported kwarg
+        with self.assertRaises(ValueError):
+            checkpoint(m0, foo="bar")
```
If you have the extra motivation, you could use `assertRaisesRegex` to show that we are raising the right error message 😃

sanketpurandare replied:
I have the motivation 💪

awgu commented on test/distributed/_composable/test_checkpoint.py:
```diff
@@ -219,6 +221,112 @@ def forward(self, x):
+        with self.assertRaises(CheckpointError):
+            m5(x.clone()).sum().backward()
+
+        # Test preserving random state
```
Should we check that running with `preserve_rng_state=True` matches without `checkpoint`?

sanketpurandare replied:
Yeah makes sense!

awgu reviewed (commented):
This looks good overall. I just left some minor comments.

sanketpurandare committed:
Update on "Adding kwargs to composable AC API to enable full capabilities"


Summary:
Firstly, this does not change any existing behaviour, since all the
default values for kwargs were hardcoded into the ``_checkpoint_without_reentrant_generator`` call.

Secondly, this is needed for unlocking the full potential of composable
checkpointing making it equivalent to ``torch.utils.checkpoint.checkpoint(use_reentrant=False)``.

Finally, an added benefit is now composable checkpointing can be used under ``FakeTensorMode`` by
passing ``preserve_rng_state=False``.

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

awgu reviewed (approved):
Looks great! 

Some lint fixes required though.

sanketpurandare committed:
Update on "Adding kwargs to composable AC API to enable full capabilities"


Summary:
Firstly, this does not change any existing behaviour, since all the
default values for kwargs were hardcoded into the ``_checkpoint_without_reentrant_generator`` call.

Secondly, this is needed for unlocking the full potential of composable
checkpointing making it equivalent to ``torch.utils.checkpoint.checkpoint(use_reentrant=False)``.

Finally, an added benefit is now composable checkpointing can be used under ``FakeTensorMode`` by
passing ``preserve_rng_state=False``.

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

sanketpurandare committed:
Update on "Adding kwargs to composable AC API to enable full capabilities"


Summary:
Firstly, this does not change any existing behaviour, since all the
default values for kwargs were hardcoded into the ``_checkpoint_without_reentrant_generator`` call.

Secondly, this is needed for unlocking the full potential of composable
checkpointing making it equivalent to ``torch.utils.checkpoint.checkpoint(use_reentrant=False)``.

Finally, an added benefit is now composable checkpointing can be used under ``FakeTensorMode`` by
passing ``preserve_rng_state=False``.

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

sanketpurandare commented:
@pytorchbot merge

sanketpurandare commented:
@pytorchbot merge
